### ISSUE TITLE
read_csv invalidates DuckDB instance

### ISSUE THREAD

**Issue by kirankpandi** [duckdb/duckdb#12314]
### What happens?

Using `read_csv` to read the content of a CSV file, attached to this ticket, returns the error - `CSV Buffer is not available to reconstruct CSV Line, please open an issue with your query and dataset.`. From this point, any query run fails with the error - `Failed: database has been invalidated because of a previous fatal error. The database must be restarted prior to being used again.`.
[sample_data.csv.gz](https://github.com/duckdb/duckdb/files/15488742/sample_data.csv.gz)
Note that we did not notice this issue in version < 0.10.2. In those versions, `read_csv` returned row length exceeded error for this file and we could successfully run other queries even after this error. 

### To Reproduce

[sample_data.csv.gz](https://github.com/duckdb/duckdb/files/15488763/sample_data.csv.gz)

1. Download the attached CSV file.
2. In DuckDB CLI, run the query `select * from read_csv('sample_data.csv.gz', HEADER=true, COMPRESSION=gzip, ALLOW_QUOTED_NULLS=false, PARALLEL=false);`. Use correct local path of the CSV file in this query.
3. Note that this query returns the error - `INTERNAL Error: CSV Buffer is not available to reconstruct CSV Line, please open an issue with your query and dataset.`
4. Now run query `select 1`. This query fails with the error - `FATAL Error: Failed: database has been invalidated because of a previous fatal error. The database must be restarted prior to being used again.
Original error: "CSV Buffer is not available to reconstruct CSV Line, please open an issue with your query and dataset."`
5. Any query would return the same error from this point on.

### OS:

MacOS

### DuckDB Version:

0.10.2, 0.10.3

### DuckDB Client:

CLI, Java

### Full Name:

Kiran Pandi

### Affiliation:

Thoughtspot

### What is the latest build you tested with? If possible, we recommend testing with the latest nightly build.

I have tested with a stable release

### Did you include all relevant data sets for reproducing the issue?

Yes

### Did you include all code required to reproduce the issue?

- [X] Yes, I have

### Did you include all relevant configuration (e.g., CPU architecture, Python version, Linux distribution) to reproduce the issue?

- [X] Yes, I have

**Comment by pdet** [comment]
> I'm getting the same error when using the python API. I think this may have something to do with the `null_padding`?
> 
> ```
> data = conn.read_csv(
>         path_or_buffer=sam_file,
>         sep="\t",
>         header=False,
>         names=list(column_names.keys()),
>         dtype=column_names,
>         sample_size=-1,
>         null_padding=True,
>     )
> ```
> 
> Error message:
> 
> ```
> InternalException: INTERNAL Error: CSV Buffer is not available to reconstruct CSV Line, please open an issue with your query and dataset.
> This error signals an assertion failure within DuckDB. This usually occurs due to unexpected conditions or errors in the program's logic.
> For more information, see https://duckdb.org/docs/dev/internal_errors
> ```

Hi Sunitj, could you also share the data that caused this error for you? The OPs error was related to a bug in single-threaded execution. Since your query is not using `PARALLEL=false` and seems to be a single file, it is possible that is something entirely different.

**Comment by sunitj** [comment]
Hello @pdet , 

Sure!

The actual data was about 60GB tab separated table. Note that the table is not exactly rectangular, some rows have more columns than others with the  columns 0-10 being consistent and column 11 onwards vary. I tried to go around this by providing the headers for the longest row and setting `null_padding=True`. I had hoped to combine the columns 11-26 into a new list column and drop 11-26 at a later step. _If there is a better way to accomplish this, please let me know._

I'm attaching two samples of the original file, please decompress before you test them:
- [duckdb_binderException.sam.gz](https://github.com/duckdb/duckdb/files/15503814/duckdb_binderException.sam.gz) contains the first 2450 rows, last two rows have more columns that all the previous rows. Results in a `BinderException`.
- [duckdb_internalError.sam.gz](https://github.com/duckdb/duckdb/files/15503816/duckdb_internalError.sam.gz) contains the first 100k rows, results in an `internalError`.

Python code to reproduce the errors:

```python
import duckdb

conn = duckdb.connect(database=":memory:")
# sam_file = "../data/20240411_DS026_C06.sortedByCoord.sam" # original file ~60GB
sam_file = "../data/duckdb_binderException.sam" # sample 1 ~ 2450 rows
# sam_file = "../data/duckdb_internalError.sam" # sample 2 ~ 100k rows
column_names = {
    "QNAME": "VARCHAR",    # NS500193:21:HWJHYBGXV:2:11102:15912:12293
    "FLAG": "USMALLINT",   # 355
    "RNAME": "VARCHAR",    # A-fermentans-SH0001780_Node_0
    "POS": "UINTEGER",     # 378385
    "MAPQ": "USMALLINT",   # 255
    "CIGAR": "VARCHAR",    # 63M
    "RNEXT": "VARCHAR",    # =
    "PNEXT": "UINTEGER",   # 378385
    "TLEN": "USMALLINT",   # 63
    "SEQ": "VARCHAR",      # AATTGGAGGCGGCACCCAGATTTGAACTGGGGATAAAGGTTTTGCAGACCTCTGCCTTACCAC
    "QUAL": "VARCHAR",     # AAAAAEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE
}
for i in range(1, 16):
    column_names[f"TAG_{i}"] = "VARCHAR"

data = conn.read_csv(
    path_or_buffer=sam_file,
    sep="\t",
    header=False,
    names=list(column_names.keys()),
    dtype=column_names,
    sample_size=-1,
    null_padding=True,
)

data.show(max_rows=5)
```

Thank you!

--
duckdb version: 0.10.3

**Comment by pdet** [comment]
> ```python
> USMALLINT
> ```

Hi @sunitj thanks for sharing the data.

I haven't been able to reproduce the internal error with latest feature, but indeed the binder error exists.

The binder error is mainly saying that the sniffer has not managed to auto-detect the options. I think the error message there could be more helpful, so I'll change that. The reason that the sniffer is unsuccessful is because we utilize only a small portion of the data to sniff column amounts, this is a known limitation and it should be fixed to use whatever is set in the `sample_size`.

```sql
FROM read_csv('/Users/holanda/Desktop/duckdb_internalError.sam',columns= {'QNAME': 'VARCHAR', 'FLAG': 'USMALLINT', 'RNAME': 'VARCHAR', 'POS': 'UINTEGER', 'MAPQ': 'USMALLINT', 'CIGAR': 'VARCHAR', 'RNEXT': 'VARCHAR', 'PNEXT': 'UINTEGER', 'TLEN': 'INTEGER', 'SEQ': 'VARCHAR', 'QUAL': 'VARCHAR', 'TAG_1': 'VARCHAR', 'TAG_2': 'VARCHAR', 'TAG_3': 'VARCHAR', 'TAG_4': 'VARCHAR', 'TAG_5': 'VARCHAR', 'TAG_6': 'VARCHAR', 'TAG_7': 'VARCHAR', 'TAG_8': 'VARCHAR', 'TAG_9': 'VARCHAR', 'TAG_10': 'VARCHAR', 'TAG_11': 'VARCHAR', 'TAG_12': 'VARCHAR', 'TAG_13': 'VARCHAR', 'TAG_14': 'VARCHAR', 'TAG_15': 'VARCHAR'}, auto_detect = false, delim = '\t', null_padding = true, header = 0)
```
With this query I was able to execute your data. Note that since you are defining the whole schema you just need to use the `columns` option with the schema struct and set `auto_detect` to false . I've also had to change `'TLEN': 'INTEGER'`, because there are some negative values in your data. Let me know if that helps :-)

**Comment by pdet** [comment]
> > ```python
> > USMALLINT
> > ```
> 
> Hi @sunitj thanks for sharing the data.
> 
> I haven't been able to reproduce the internal error with latest feature, but indeed the binder error exists.
> 
> The binder error is mainly saying that the sniffer has not managed to auto-detect the options. I think the error message there could be more helpful, so I'll change that. The reason that the sniffer is unsuccessful is because we utilize only a small portion of the data to sniff column amounts, this is a known limitation and it should be fixed to use whatever is set in the `sample_size`.
> 
> ```sql
> FROM read_csv('/Users/holanda/Desktop/duckdb_internalError.sam',columns= {'QNAME': 'VARCHAR', 'FLAG': 'USMALLINT', 'RNAME': 'VARCHAR', 'POS': 'UINTEGER', 'MAPQ': 'USMALLINT', 'CIGAR': 'VARCHAR', 'RNEXT': 'VARCHAR', 'PNEXT': 'UINTEGER', 'TLEN': 'INTEGER', 'SEQ': 'VARCHAR', 'QUAL': 'VARCHAR', 'TAG_1': 'VARCHAR', 'TAG_2': 'VARCHAR', 'TAG_3': 'VARCHAR', 'TAG_4': 'VARCHAR', 'TAG_5': 'VARCHAR', 'TAG_6': 'VARCHAR', 'TAG_7': 'VARCHAR', 'TAG_8': 'VARCHAR', 'TAG_9': 'VARCHAR', 'TAG_10': 'VARCHAR', 'TAG_11': 'VARCHAR', 'TAG_12': 'VARCHAR', 'TAG_13': 'VARCHAR', 'TAG_14': 'VARCHAR', 'TAG_15': 'VARCHAR'}, auto_detect = false, delim = '\t', null_padding = true, header = 0)
> ```
> 
> With this query I was able to execute your data. Note that since you are defining the whole schema you just need to use the `columns` option with the schema struct and set `auto_detect` to false . I've also had to change `'TLEN': 'INTEGER'`, because there are some negative values in your data. Let me know if that helps :-)

I've managed to reproduce the internal error with the query
```sql 
I've managed to reproduce the internal error with:
FROM read_csv('/Users/holanda/Desktop/duckdb_internalError.sam', names = ['QNAME', 'FLAG', 'RNAME', 'POS', 'MAPQ', 'CIGAR', 'RNEXT', 'PNEXT', 'TLEN', 'SEQ', 'QUAL', 'TAG_1', 'TAG_2', 'TAG_3', 'TAG_4', 'TAG_5', 'TAG_6', 'TAG_7', 'TAG_8', 'TAG_9', 'TAG_10', 'TAG_11', 'TAG_12', 'TAG_13', 'TAG_14', 'TAG_15'], types= {'QNAME': 'VARCHAR', 'FLAG': 'USMALLINT', 'RNAME': 'VARCHAR', 'POS': 'UINTEGER', 'MAPQ': 'USMALLINT', 'CIGAR': 'VARCHAR', 'RNEXT': 'VARCHAR', 'PNEXT': 'UINTEGER', 'TLEN': 'USMALLINT', 'SEQ': 'VARCHAR', 'QUAL': 'VARCHAR', 'TAG_1': 'VARCHAR', 'TAG_2': 'VARCHAR', 'TAG_3': 'VARCHAR', 'TAG_4': 'VARCHAR', 'TAG_5': 'VARCHAR', 'TAG_6': 'VARCHAR', 'TAG_7': 'VARCHAR', 'TAG_8': 'VARCHAR', 'TAG_9': 'VARCHAR', 'TAG_10': 'VARCHAR', 'TAG_11': 'VARCHAR', 'TAG_12': 'VARCHAR', 'TAG_13': 'VARCHAR', 'TAG_14': 'VARCHAR', 'TAG_15': 'VARCHAR'}, sample_size=-1,null_padding=True)
```
on a branch prior to the #12315 fix got merged. So turns out that it was the same issue in the OP, the problem would happen during sniffing where the scanning runs single-threaded :-), so that got fixed!

**Comment by Mytherin** [comment]
This should be fixed now by https://github.com/duckdb/duckdb/pull/12315

**Comment by kirankpandi** [comment]
@Mytherin I see that https://github.com/duckdb/duckdb/pull/12315 is not included in DuckDB 1.0. Do you have an ETA for the next release that might include this fix ?

**Comment by Mytherin** [comment]
It should be in 1.1 which is currently schedule to be released in 3 months
